sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -0,0 +1,488 @@
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.apache.beam.sdk.values.PInput;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+import javax.annotation.Nullable;
+import javax.jms.Connection;
+import javax.jms.ConnectionFactory;
+import javax.jms.Destination;
+import javax.jms.MessageConsumer;
+import javax.jms.MessageProducer;
+import javax.jms.Session;
+import javax.jms.TextMessage;
+
+/**
+ * An unbounded source for JMS destinations (queues or topics).
+ *
+ * <h3>Reading from a JMS destination</h3>
+ *
+ * <p>JmsIO source returns unbounded collection of JMS records as {@code PCollection<JmsRecord<T>>}.
+ * A {@link JmsRecord} includes JMS headers and properties, along with the JMS message payload.</p>
+ *
+ * <p>To configure a JMS source, you have to provide a {@code javax.jms.ConnectionFactory}
+ * and the destination (queue or topic) where to consume. The following example
+ * illustrates various options for configuring the source:</p>
+ *
+ * <pre>{@code
+ *
+ * pipeline.apply(JmsIO.read()
+ *    .withConnectionFactory(myConnectionFactory)
+ *    .withQueue("my-queue")
+ *    // above two are required configuration, returns PCollection<JmsRecord<byte[]>>
+ *
+ *    // rest of the settings are optional
+ *
+ * }</pre>
+ *
+ * <h3>Writing to a JMS destination</h3>
+ *
+ * JmsIO sink supports writing test messages to a JMS destination on a broker.
+ * To configure a JMS sink, you must specify a {@link javax.jms.ConnectionFactory} and a
+ * {@link javax.jms.Destination} name.
+ * For instance:
+ *
+ * <pre>{@code
+ *
+ *  pipeline
+ *    .apply(...) // returns PCollection<String>
+ *    .apply(JmsIO.write()
+ *        .withConnectionFactory(myConnectionFactory)
+ *        .withQueue("my-queue")
+ *
+ * }</pre>
+ */
+public class JmsIO {
+
+  private static final Logger LOG = LoggerFactory.getLogger(JmsIO.class);
+
+  public static Read read() {
+    return new Read();
+  }
+
+  public static Write write() {
+    return new Write();
+  }
+
+  /**
+   * A {@link PTransform} to read from a JMS destination. See {@link JmsIO} for more
+   * information on usage and configuration.
+   */
+  public static class Read extends PTransform<PBegin, PCollection<JmsRecord>> {
+
+    public Read withConnectionFactory(ConnectionFactory connectionFactory) {
+      return new Read(connectionFactory, queue, topic, maxNumRecords, maxReadTime);
```

Comment:
totally optional: do you wish to do any validation here and below? e.g.,

* `checkNotNull(connectionFactory, "connectionFactory")`
* `checkArgument(maxNumRecords >= 0)`